process/block/preprocess/transactionsV2.go

```diff
@@ -738,3 +488,313 @@ func getAdditionalTimeFunc() func() bool {
+			"hash", txHash,
+			"sender shard", senderShardID,
+			"receiver shard", receiverShardID)
+		return nil, nil, false
+	}
+
+	if len(cpmbi.senderAddressToSkip) > 0 {
+		if bytes.Equal(cpmbi.senderAddressToSkip, tx.GetSndAddr()) {
+			cpmbi.processingInfo.numTxsSkipped++
+			return nil, nil, false
+		}
+	}
+
+	miniBlock, ok := cpmbi.mapMiniBlocks[receiverShardID]
+	if !ok {
+		log.Debug("mini block is not created", "shard", receiverShardID)
+		return nil, nil, false
+	}
+
+	addressHasEnoughBalance := txs.hasAddressEnoughInitialBalance(tx)
+	if !addressHasEnoughBalance {
+		cpmbi.processingInfo.numTxsWithInitialBalanceConsumed++
+		return nil, nil, false
+	}
+
+	return tx, miniBlock, true
+}
+
+func (txs *transactions) getTxAndMbInfo(
+	tx *transaction.Transaction,
+	isMiniBlockEmpty bool,
+	receiverShardID uint32,
+	cpmbi *createAndProcessMiniBlocksInfo,
+) *txAndMbInfo {
+	_, txTypeDstShard := txs.txTypeHandler.ComputeTransactionType(tx)
+	isReceiverSmartContractAddress := txTypeDstShard == process.SCDeployment || txTypeDstShard == process.SCInvoking
+	scTxType := nonScTx
+	if isReceiverSmartContractAddress {
+		scTxType = scTx
+	}
+
+	numNewMiniBlocks := 0
+	if isMiniBlockEmpty || cpmbi.firstCrossShardScCallOrSpecialTxFound {
+		numNewMiniBlocks = 1
+	}
+	numNewTxs := 1
+
+	isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
+		(isReceiverSmartContractAddress || len(tx.RcvUserName) > 0)
+	if isCrossShardScCallOrSpecialTx {
+		if !cpmbi.firstCrossShardScCallOrSpecialTxFound {
+			numNewMiniBlocks++
+		}
+		if cpmbi.mapCrossShardScCallsOrSpecialTxs[receiverShardID] >= cpmbi.maxCrossShardScCallsOrSpecialTxsPerShard {
+			numNewTxs += core.AdditionalScrForEachScCallOrSpecialTx
+		}
+	}
+
+	return &txAndMbInfo{
+		numNewTxs:                      numNewTxs,
+		numNewMiniBlocks:               numNewMiniBlocks,
+		isReceiverSmartContractAddress: isReceiverSmartContractAddress,
+		isCrossShardScCallOrSpecialTx:  isCrossShardScCallOrSpecialTx,
+		txType:                         scTxType,
+	}
+}
+
+func (txs *transactions) applyExecutedTransaction(
+	tx *transaction.Transaction,
+	txHash []byte,
+	miniBlock *block.MiniBlock,
+	receiverShardID uint32,
+	txMbInfo *txAndMbInfo,
+	cpmbi *createAndProcessMiniBlocksInfo,
+) {
+	cpmbi.senderAddressToSkip = []byte("")
+
+	if txs.balanceComputation.IsAddressSet(tx.GetSndAddr()) {
+		txMaxTotalCost := txs.getTxMaxTotalCost(tx)
+		ok := txs.balanceComputation.SubBalanceFromAddress(tx.GetSndAddr(), txMaxTotalCost)
+		if !ok {
+			log.Error("applyExecutedTransaction.SubBalanceFromAddress",
+				"sender address", tx.GetSndAddr(),
+				"tx max total cost", txMaxTotalCost,
+				"err", process.ErrInsufficientFunds)
+		}
+	}
+
+	isMiniBlockEmpty := len(miniBlock.TxHashes) == 0
+	firstMiniBlockSplitForReceiverShardFound := (txMbInfo.isReceiverSmartContractAddress && cpmbi.mapScsForShard[receiverShardID] == 0 && cpmbi.mapTxsForShard[receiverShardID] > 0) ||
```

Comment:
this is an incredibly complicated condition. can we make it easier ?

Reply:
refactored